physutils/io.py

```diff
@@ -0,0 +1,223 @@
+        )
+    logger.info(f"Saved {data} in {fname}")
+
+    return fname
+
+
+def load_history(file, verbose=False):
+    """
+    Loads history from `file` and replays it, creating new Physio instance
+
+    Parameters
+    ----------
+    file : str
+        Path to input JSON file
+    verbose : bool, optional
+        Whether to print messages as history is being replayed. Default: False
+
+    Returns
+    -------
+    file : str
+        Full filepath to saved output
+    """
+
+    # import inside function for safety!
+    # we'll likely be replaying some functions from within this module...
+    import physutils
+
+    # grab history from provided JSON file
+    with open(file, "r") as src:
+        history = json.load(src)
+
+    # replay history from beginning and return resultant Physio object
+    logger.info(f"Replaying history from {file}")
+    data = None
+    for func, kwargs in history:
+        if verbose:
+            logger.info("Rerunning {}".format(func))
+        # loading functions don't have `data` input because it should be the
+        # first thing in `history` (when the data was originally loaded!).
+        # for safety, check if `data` is None; someone could have potentially
+        # called load_physio on a Physio object (which is a valid, albeit
+        # confusing, thing to do)
+        if "load" in func and data is None:
+            if not op.exists(kwargs["data"]):
+                if kwargs["data"].startswith("/"):
+                    msg = (
+                        "Perhaps you are trying to load a history file "
+                        "that was generated with an absolute path?"
+                    )
+                else:
+                    msg = (
+                        "Perhaps you are trying to load a history file "
+                        "that was generated from a different directory?"
+                    )
+                raise FileNotFoundError(
+                    "{} does not exist. {}".format(kwargs["data"], msg)
+                )
+            data = getattr(physutils, func)(**kwargs)
+        else:
+            data = getattr(physutils, func)(data, **kwargs)
+
+    return data
+
+
+def save_history(file, data):
+    """
+    Saves history of physiological `data` to `file`
+
+    Saved file can be replayed with `peakdet.load_history`
+
+    Parameters
+    ----------
+    file : str
+        Path to output file; .json will be appended if necessary
+    data : Physio_like
+        Data with history to be saved to file
+
+    Returns
+    -------
+    file : str
+        Full filepath to saved output
+    """
+
+    from physutils.physio import check_physio
+
+    data = check_physio(data)
+    if len(data.history) == 0:
+        logger.warning(
+            "History of provided Physio object is empty. Saving "
+            "anyway, but reloading this file will result in an "
+            "error."
```

Comment:
Loading an empty history file does not lead to an error, but just to `None`. Maybe we can just rephrase the warning message here, so that it is more clear ?

Reply:
Would it be more useful to throw an error when loading an empty history instead?

Reply:
Are there any instances where we would want to load an empty history? Off-hand, I think a warning is preferable for flexibility unless we're absolutely sure we want this to be an error.